### PR TITLE
Look up `dub.selections.json` files in parent directories too

### PR DESCRIPTION
**Pull request by kinke**
In case the root package directory doesn't contain a `dub.selections.json` file, dub now looks in parent directories too and potentially uses the first (deepest) one it finds - iff that JSON file contains an optional new `"inheritable": true` flag.

This allows using a 'central' `dub.selections.json` file for a repository containing multiple dub projects, making it automatically apply to all builds in that source tree if located in the repository root directory (unless a local `dub.selections.json` overrides it).

Such an inherited selections file is never mutated when running dub for a nested project, i.e., changes are always saved to a *local* `dub.selections.json` file.

### CODE REVIEW

**Comment by kinke** [comment]
Ouch, looks like the testsuite fails due to dub's own `dub.selections.json` in its repo root...

**Comment by WebFreak001** [comment]
I think it's probably fine to add the feature, but I'm not sure yet on the semantics. If we just inherit from parent directories that would be nice for sub-packages of projects, which want to read from the dub.selections.json anyway and not need to manually `dub upgrade` all the sub-packages.

I think it might be best to have some kind of `"inheritSelections": true` property in the dub.json/sdl, but I'm not yet fully set on that thought.

What do you think @s-ludwig ?

**Comment by Geod24** [comment]
[To quote](https://github.com/dlang/dub/issues/1881#issuecomment-588397010):

> The general idea with sub packages is that they are logically completely separate packages, with the exception of being tied to the same version. Everything in addition to that would lead to another dimension of opt-in/opt-out directives that become necessary to override the default inheritance behavior.

I think we're ought to either stick with this principle and properly document it, OR re-evaluate what is people's natural expectation / need around subpackages. It's confusing even to maintainers when subpackages should be used vs configurations.

**Comment by nordlow** [comment]
How do other languages solves these issues? Would be interesting to check if Rust have these same issues and implement them in a similar way to what we plan.

**Comment by nordlow** [comment]
Apart from merging with master what remains in this MR, @kinke?

**Comment by kinke** [comment]
> Apart from merging with master what more work needs to happen in this MR?

Well, it's a pretty big breaking change and might not be always desirable. So maybe make it opt-in via cmdline flag or extending the dub.selections.json file (`inheritable: true` or so - but that would still lead to more file lookups).

> What is meant by "changes are always saved to a local dub.selections.json file."?

If you run e.g. `dub upgrade` in a nested project, changes are saved to a dub.selections.json file *in that dir*, not to the inherited file from some parent dir.

File: source/dub/project.d
Comment (WebFreak001) [inline source/dub/project.d:1766]:
```
source/dub/project.d(1766,5): Deprecation: function `dub.dependency.Dependency.path` is deprecated - Construct a new `Dependency` object instead
```

Comment (kinke) [reply source/dub/project.d]: Yeah thx, seen it via that awesome bot too, cool stuff! :+1:

**Comment by kinke** [comment]
Rebased and made opt-in via a new optional `inheritable: true` in the `dub.selections.json` file to be inherited by nested projects.

I've left the `fileVersion` at 1 for now; I guess we'd have to bump it. But maybe only for files containing `inheritable: true`, so that the whole D ecosystem doesn't have to be adapted.

**Comment by WebFreak001** [comment]
I think keeping fileVersion at 1 is fine here, since it's not backwards incompatible

**Comment by kinke** [comment]
Ping.

File: source/dub/project.d
Comment (WebFreak001) [inline source/dub/project.d:2011]: in case of invalid selections files, shouldn't we rather error than silently ignoring the file?

**Review by WebFreak001 — CHANGES_REQUESTED** [review]

File: changelog/selections_from_parent_dir.dd
Comment (kinke) [reply changelog/selections_from_parent_dir.dd:6]: It wouldn't make it clear that for some project in `root/dir1/dir2`, an inheritable `root/dub.selections.json` wouldn't apply if there's a non-inheritable `root/dir1/dub.selections.json`. `iff` should be pretty common in IT specs. :)

File: source/dub/project.d
Comment (kinke) [reply source/dub/project.d:2011]: I'm pretty sure I kept the behavior as-is (at that time - merge conflicts now obviously after all that time) - an empty `Selections` object if there's no file or the file is invalid.

**Comment by atilaneves** [comment]
Given the opt-in nature now with having to set inheritability in the parent selections file, I don't see any reason to not merget his. Objections?

**Comment by kinke** [comment]
Updated yet again to resolve merge conflicts. The change is better integrated now; I got rid of the earlier new extra helper (`lookupAndParseSelectionsFile()` or so), and instead adapted the new (in master) `PackageManager.readSelections()` method, incl. making that `public`. The reason is that reggae needs to know the path to the used `dub.selections.json`, so that it can add a proper dependency in the generated build scripts, leading to a reggae re-run whenever a used `dub.selections.json` has changed.

File: test/selections-from-parent-dir.sh
Comment (Geod24) [inline test/selections-from-parent-dir.sh:10]: Any chance you could convert this to the new test framework ? E.g. see `source/dub/test/other.d`.

File: changelog/selections_from_parent_dir.dd
Comment (Geod24) [inline changelog/selections_from_parent_dir.dd:16]: Can we add a test for this ?

File: source/dub/packagemanager.d
Comment (kinke) [reply source/dub/packagemanager.d:67]: 2. would be doable after a quick glance, but would need a changed signature for the public `SelectionsFile.fromYAML()` factory function (unused in dub AFAICT), adding a `path` parameter.

Comment (kinke) [reply source/dub/packagemanager.d]: Yeah I figured, so I went with option 1.

File: source/dub/packagemanager.d
Comment (kinke) [reply source/dub/packagemanager.d]: Non-final now.

File: test/selections-from-parent-dir.sh
Comment (kinke) [reply test/selections-from-parent-dir.sh]: I did; pretty cool. :+1: - `dub test -c library-nonet` fails for some rather bizarre reason (the CI problem), `dub test` alone works fine.

Comment (kinke) [reply test/selections-from-parent-dir.sh]: Solved; was a problem of the mocked file system (no entries for `.`  and `..`, so needs extra normalization; might be required in other places too).

File: source/dub/packagemanager.d
Comment (Geod24) [inline source/dub/packagemanager.d:1183]: Logically this would be better represented as something that returns a path that is then acted on. The code as-is does not mean to be changed (it works, that's what matters) but a loop that ends up with an unconditional `break` is not a loop (or IOW, contains too much code).

File: source/dub/project.d
Comment (Geod24) [inline source/dub/project.d:111]: How do we know here that the path is absolute ?
Loops like we should just change the signature of the function ?

File: source/dub/recipe/selection.d
Comment (Geod24) [inline source/dub/recipe/selection.d:58]: I don't know if we should expose this convenience function here. The intent of `SelectionFile` is just to wrap a specific `Selections!uint`. However, the implementation makes sense logically. IMO the question boils down to: "Do we ever find ourselves in a situation where we manipulate an invalid `Selections!uint`?". I would be tempted to say no, as trying to make a generic `Dependency` instead of the discriminated union it is now is what made the code a mess.

File: source/dub/packagemanager.d
Comment (Geod24) [inline source/dub/packagemanager.d:71]: I would like to fold this into `SelectionsFile`, but I know it's not currently possible. All things considered, it might be a good thing to add to `Configy`, so I'll have a look at it myself. It should not block this being merged.

File: source/dub/test/selections_from_parent_dir.d
Comment (Geod24) [inline source/dub/test/selections_from_parent_dir.d:1]: Module doc please

File: source/dub/test/selections_from_parent_dir.d
Comment (Geod24) [inline source/dub/test/selections_from_parent_dir.d:48]: `readText`

File: source/dub/test/selections_from_parent_dir.d
Comment (Geod24) [inline source/dub/test/selections_from_parent_dir.d:112]: Maybe those tests should check the version too ? IIUC before the `rm` `hasAlldependencies` should be `false`, and it should be `true` afterwards. You can use `TestDub.newTest` to easily test this.

**Review by Geod24 — APPROVED** [review]
Looks good! Made a few suggestions. I thing there's nothing blocking the merge (although I would like to see the points addressed, especially the ones about `inheritable` property and absolute path, as that could trigger an assert easily) but can do it myself if you just want to go agead.

**Comment by Geod24** [comment]
@kinke : I took the liberty to push the requested fixes, as I know you are away and I'm sure you'd like this fixed sooner than later.

**Comment by Geod24** [comment]
> This is a scheduled macOS-11 brownout. The macOS-11 environment is deprecated and will be removed on June 28th, 2024.
 
 Well...

**Comment by Geod24** [comment]
```
../../../.dub/packages/vibe-core/2.7.1/vibe-core/source/vibe/core/path.d(72,8): Error: none of the overloads of `endsWithSlash` are callable using argument types `(bool) const`
../../../.dub/packages/vibe-core/2.7.1/vibe-core/source/vibe/core/path.d(510,17):        Candidates are: `vibe.core.path.GenericPath!(PosixPathFormat).GenericPath.endsWithSlash() const`
../../../.dub/packages/vibe-core/2.7.1/vibe-core/source/vibe/core/path.d(512,17):                        `vibe.core.path.GenericPath!(PosixPathFormat).GenericPath.endsWithSlash(bool v)`
source/dub/project.d(127,60): Error: template instance `vibe.core.path.relativeTo!(const(GenericPath!(PosixPathFormat)))` error instantiating
```

Well looks like vibe-core has a bug too, but we can work around that.

**Comment by Geod24** [comment]
None of the workaround is pretty though, so I opened https://github.com/vibe-d/vibe-core/pull/400

**Comment by kinke** [comment]
Thanks Mathias!

**Pull request merged**
